File: 荔枝网 wasm/使用skill逆向/generate_headers.py
#!/usr/bin/env python3
"""
荔枝网 API 数据获取（cloakbrowser 版）

直接利用浏览器内已初始化的 React app + WASM 签名器：
  1. 在页面内发 fetch/XHR 请求
  2. 用 page.route 拦截签名头
  3. Python 侧重放签名头获取数据

或者更直接：在 page 内用 fetch() 发请求（走完整签名链），
用 page.route 转发响应给 Python。

用法：python generate_headers.py
"""

import json
from cloakbrowser import launch

UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36"


def fetch_api(page, path, query=""):
    """
    在浏览器页面内调用 fetch() 发 API 请求。
    由于页面已加载了 app 代码（含 axios 拦截器 + WASM 签名器），
    页面内的 fetch() 调用会自动经过签名拦截器附加 ca 头。

    我们用 page.route 拦截响应，不实际发送到外网。

    如果直接让浏览器发真实请求，page.route 可以捕获响应。
    """
    full_url = f"https://gdtv-api.gdtv.cn{path}"
    if query:
        full_url += "?" + query

    response_data = {}
    done = False

    def handle_route(route):
        nonlocal response_data, done
        if path in route.request.url:
            # 捕获请求头（用于后续 Python 重放）
            response_data["request_headers"] = dict(route.request.headers)
            # 让请求继续发送，获取真实响应
            route.continue_()

    page.route(f"**{path}**", handle_route)

    # 在页面中通过 window.fetch 发请求
    # 注意：这会经过浏览器的完整网络栈
    try:
        page.evaluate(f"""
            fetch('{full_url}', {{
                method: 'GET',
                headers: {{ 'Accept': 'application/json, text/plain, */*' }}
            }}).then(r => r.json()).then(d => {{
                window.__api_response = d;
            }}).catch(e => {{
                window.__api_error = e.message;
            }});
        """)
    except:
        pass

    # 等待响应
    page.wait_for_timeout(5000)
    page.unroute(f"**{path}**")

    api_response = page.evaluate("window.__api_response")
    api_error = page.evaluate("window.__api_error")

    return {
        "headers": response_data.get("request_headers", {}),
        "data": api_response,
        "error": api_error,
    }


def main():
    print("Launching browser...")
    b = launch(headless=True)
    page = b.new_page()

    print("Navigating to gdtv.cn...")
    page.goto("https://www.gdtv.cn/", wait_until="networkidle")
    page.wait_for_timeout(6000)  # 等 WASM 初始化

    # Test API fetch
    print("\nTesting fetch...")
    result = fetch_api(page, "/api/channel/v1/news", "beginScore=0&pageSize=3&channelId=117")

    print(f"\nStatus: {'ok' if result['data'] else 'failed'}")
    print(f"Error: {result['error']}")

    sign_headers = {k: v for k, v in result.get('headers', {}).items() if k.lower().startswith('x-itouchtv')}
    if sign_headers:
        print("Sign headers captured:")
        for k, v in sign_headers.items():
            print(f"  {k}: {v}")

    if result.get('data'):
        print(f"\nResponse preview:")
        print(json.dumps(result['data'], ensure_ascii=False, indent=2)[:500])

    page.wait_for_timeout(1000)
    b.close()


if __name__ == "__main__":
    main()

File: 荔枝网 wasm/使用skill逆向/generate_headers.js
#!/usr/bin/env node
/**
 * 荔枝网 (gdtv.cn) API 签名中介
 *
 * 原理：
 *   荔枝网使用 WASM (itouchtv_webqs) 对每个 API 请求生成 ca-signature。
 *   签名算法封装在 webpack 模块闭包中，无法直接提取。
 *
 *   本脚本利用 sdenv 加载 gdtv.cn → 初始化 WASM 签名器后，
 *   在 jsdom 内用 XMLHttpRequest 发出请求（复用已初始化的签名逻辑），
 *   将响应数据返回给 Python。
 *
 *   Python 只需指定 path + query，本脚本负责完整请求并返回 JSON body。
 *
 * 输入(stdin JSON)：
 *   {"method":"GET", "path":"/api/channel/v1/news", "query":"beginScore=0&pageSize=11&channelId=117"}
 *
 *   或批量：
 *   {"requests": [{"path":"...","query":"..."}, ...]}
 *
 * 输出(stdout JSON)：
 *   {"success":true, "results":[{"url":"...","data":{...}}, ...]}
 */

'use strict';
const { jsdomFromUrl } = require('sdenv');
const vm = require('vm');

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36';
const BASE_URL = 'https://www.gdtv.cn';
const API_BASE = 'https://gdtv-api.gdtv.cn';

function readStdin() {
  return new Promise(resolve => {
    if (process.stdin.isTTY) { resolve(null); return; }
    let data = '';
    process.stdin.setEncoding('utf-8');
    process.stdin.on('data', chunk => data += chunk);
    process.stdin.on('end', () => {
      try { resolve(JSON.parse(data)); }
      catch (e) { resolve(null); }
    });
    setTimeout(() => resolve(null), 2000);
  });
}

async function main() {
  const input = await readStdin();
  const requests = (input && input.requests) ? input.requests : [{
    method: (input && input.method) || 'GET',
    path: (input && input.path) || '/api/channel/v1/news',
    query: (input && input.query) || 'beginScore=0&pageSize=11&channelId=117',
  }];

  // Step 1: sdenv 加载 gdtv.cn 首页
  const dom = await jsdomFromUrl(BASE_URL + '/', {
    userAgent: UA,
    consoleConfig: { error: () => {} },
  });
  const { window } = dom;

  // 等待前端框架 + WASM 签名器初始化
  await new Promise(r => setTimeout(r, 8000));

  // Step 2: 在 jsdom 上下文中注入工具函数
  // 利用已初始化的 axios 实例发请求（axios 拦截器会自动附加签名头）
  const context = dom.getInternalVMContext();
  vm.runInContext(`
    window.__ad_results = [];
    window.__ad_fetch = function(url) {
      return new Promise(function(resolve, reject) {
        var xhr = new XMLHttpRequest();
        xhr.open('GET', url);
        xhr.onload = function() {
          if (xhr.status >= 200 && xhr.status < 300) {
            try {
              resolve({status: xhr.status, body: JSON.parse(xhr.responseText)});
            } catch(e) {
              resolve({status: xhr.status, body: xhr.responseText});
            }
          } else {
            resolve({status: xhr.status, error: xhr.statusText, body: xhr.responseText});
          }
        };
        xhr.onerror = function() {
          resolve({status: 0, error: 'Network error'});
        };
        xhr.timeout = 15000;
        xhr.ontimeout = function() {
          resolve({status: 0, error: 'Timeout'});
        };
        xhr.send();
      });
    };
  `, context);

  // Step 3: 依次请求
  const results = [];
  for (const r of requests) {
    const fullUrl = API_BASE + r.path + (r.query ? '?' + r.query : '');
    try {
      const result = await vm.runInContext(
        `window.__ad_fetch('${fullUrl}')`, context
      );
      results.push({ url: fullUrl, ...result });
    } catch (e) {
      results.push({ url: fullUrl, error: e.message });
    }
  }

  try { window.close(); } catch (e) {}

  process.stdout.write(JSON.stringify({
    success: true,
    results: results,
  }));

  setTimeout(() => process.exit(0), 500);
}

main().catch(err => {
  process.stdout.write(JSON.stringify({
    success: false, error: err.message, results: [],
  }));
  setTimeout(() => process.exit(1), 500);
});
